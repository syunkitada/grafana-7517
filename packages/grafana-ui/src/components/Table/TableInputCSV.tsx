import React from 'react';
import debounce from 'lodash/debounce';
import Papa, { ParseError, ParseMeta } from 'papaparse';
import { TableData, Column } from '../../types/data';

// Subset of all parse configs
export interface ParseConfig {
  delimiter?: string; // default: ","
  newline?: string; // default: "\r\n"
  quoteChar?: string; // default: '"'
  encoding?: string; // default: ""
  comments?: boolean | string; // default: false
}

interface ParseResults {
  table: TableData;
  meta: ParseMeta;
  errors: ParseError[];
}

// This mutates the table structure!
export function checkAndFix(table: TableData): number {
  let cols = table.columns.length;
  let different = 0;
  table.rows.forEach(row => {
    if (cols !== row.length) {
      different++;
      cols = Math.max(cols, row.length);
    }
  });
  if (different > 0) {
    if (cols !== table.columns.length) {
      const diff = cols - table.columns.length;
      for (let i = 0; i < diff; i++) {
        table.columns.push({
          text: 'Column ' + table.columns.length,
        });
      }
    }
    table.rows.forEach(row => {
      const diff = cols - row.length;
      for (let i = 0; i < diff; i++) {
        row.push(null);
      }
    });
  }
  return different;
}

export function parseCSV(text: string, config?: ParseConfig): ParseResults {
  const results = Papa.parse(text, { ...config, dynamicTyping: true, skipEmptyLines: true });

  const { data, meta, errors } = results;
  if (!data || data.length < 1) {
    if (!text) {
      errors.length = 0; // clear other errors
    }
    errors.push({
      type: 'warning', // A generalization of the error
      message: 'Empty Data',
      code: 'empty',
      row: 0,
    });
    return {
      table: {
        columns: [],
        rows: [],
        type: 'table',
        columnMap: {},
      } as TableData,
      meta,
      errors,
    };
  }

  const first = results.data.shift();
  const table = {
    columns: first.map((v: any, index: number) => {
      if (!v) {
        v = 'Column ' + (index + 1);
      }
      return {
        text: v.toString().trim(),
      } as Column;
    }),
    rows: results.data,
    type: 'table',
    columnMap: {},
  } as TableData;

  const changed = checkAndFix(table);
  if (changed > 0) {
    errors.push({
      type: 'warning', // A generalization of the error
      message: 'not all rows have the same width. Changed:' + changed,
      code: 'width',
      row: 0,
    });
  }
  return {
    table,
    meta,
    errors,
  };
}

interface Props {
  config?: ParseConfig;
  width: number;
  height: number;
}

interface State {
  text: string;
  results: ParseResults;
}

class TableInputCSV extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      text: '',
      results: parseCSV('', this.props.config),
    };
  }

  readCSV = debounce(() => {
    const results = parseCSV(this.state.text, this.props.config);
    this.setState({ results });
    console.log('GOT:', results);
  }, 150);

  componentDidUpdate(prevProps: Props, prevState: State) {
    if (this.state.text !== prevState.text || this.props.config !== prevProps.config) {
      this.readCSV();
    }
  }

  handleChange = (event: any) => {
    this.setState({ text: event.target.value });
  };
  handleBlur = (event: React.SyntheticEvent<HTMLTextAreaElement>) => {
    //  console.log('BLUR', event);
  };

  render() {
    const { width, height } = this.props;
    const { table, errors } = this.state.results;

    return (
      <div
        className={'gf-table-input-wrap'}
        style={{
          width: `${width}px`,
          height: `${height}px`,
        }}
      >
        <textarea value={this.state.text} onChange={this.handleChange} onBlur={this.handleBlur} />
        <footer>
          BAR: / ROWS:{table.rows.length} / COLS:{table.columns.length} / {JSON.stringify(errors)}
        </footer>
      </div>
    );
  }
}

export default TableInputCSV;
