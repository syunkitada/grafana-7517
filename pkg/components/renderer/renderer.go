package renderer

import (
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"time"

	"strconv"

	"github.com/grafana/grafana/pkg/log"
	"github.com/grafana/grafana/pkg/middleware"
	"github.com/grafana/grafana/pkg/setting"
	"github.com/grafana/grafana/pkg/util"
	"strings"
)

type RenderOpts struct {
	Path       string
	Width      string
	Height     string
	Timeout    string
	OrgId      int64
	TimeOffset string
}

var rendererLog log.Logger = log.New("png-renderer")

func isoTimeOffsetToPosixTz(isoOffset string) string {
	re := regexp.MustCompile(`^([+-])([0-1][0-9]|2[0-4])([0-5][0-9])$`)
	results := re.FindStringSubmatch(isoOffset)
	if results == nil {
		return ""
	}
	sign := "+"
	if results[1] == "+" {
		sign = "-"  // "+" is west and "-" is east in POSIX TZ
	}
	return fmt.Sprintf("SOMEWHERE%s%s:%s", sign, results[2], results[3])
}

func appendEnviron(baseEnviron []string, name string, value string) []string {
	results := make([]string, 0)
	prefix := fmt.Sprintf("%s=", name)
	for _, v := range baseEnviron {
		if !strings.HasPrefix(v, prefix) {
			results = append(results, v)
		}
	}
	return append(results, fmt.Sprintf("%s=%s", name, value))
}

func RenderToPng(params *RenderOpts) (string, error) {
	rendererLog.Info("Rendering", "path", params.Path)

	var executable = "phantomjs"
	if runtime.GOOS == "windows" {
		executable = executable + ".exe"
	}

	localDomain := "localhost"
	if setting.HttpAddr != setting.DEFAULT_HTTP_ADDR {
		localDomain = setting.HttpAddr
	}

	url := fmt.Sprintf("%s://%s:%s/%s", setting.Protocol, localDomain, setting.HttpPort, params.Path)

	binPath, _ := filepath.Abs(filepath.Join(setting.PhantomDir, executable))
	scriptPath, _ := filepath.Abs(filepath.Join(setting.PhantomDir, "render.js"))
	pngPath, _ := filepath.Abs(filepath.Join(setting.ImagesDir, util.GetRandomString(20)))
	pngPath = pngPath + ".png"

	renderKey := middleware.AddRenderAuthKey(params.OrgId)
	defer middleware.RemoveRenderAuthKey(renderKey)

	cmdArgs := []string{
		"--ignore-ssl-errors=true",
		"--web-security=false",
		scriptPath,
		"url=" + url,
		"width=" + params.Width,
		"height=" + params.Height,
		"png=" + pngPath,
		"domain=" + localDomain,
		"renderKey=" + renderKey,
	}

	cmd := exec.Command(binPath, cmdArgs...)
	stdout, err := cmd.StdoutPipe()

	if err != nil {
		return "", err
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		return "", err
	}

	tz := isoTimeOffsetToPosixTz(params.TimeOffset)
	if tz != "" {
		baseEnviron := os.Environ()
		cmd.Env = appendEnviron(baseEnviron, "TZ", tz)
	}

	err = cmd.Start()
	if err != nil {
		return "", err
	}

	go io.Copy(os.Stdout, stdout)
	go io.Copy(os.Stdout, stderr)

	done := make(chan error)
	go func() {
		cmd.Wait()
		close(done)
	}()

	timeout, err := strconv.Atoi(params.Timeout)
	if err != nil {
		timeout = 15
	}

	select {
	case <-time.After(time.Duration(timeout) * time.Second):
		if err := cmd.Process.Kill(); err != nil {
			rendererLog.Error("failed to kill", "error", err)
		}
		return "", fmt.Errorf("PhantomRenderer::renderToPng timeout (>%vs)", timeout)
	case <-done:
	}

	rendererLog.Debug("Image rendered", "path", pngPath)
	return pngPath, nil
}
