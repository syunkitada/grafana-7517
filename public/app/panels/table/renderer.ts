
export class TableRenderer {
  constructor(private panel, private table) {
  }

  formatColumnValue(columnIndex, value) {
    return "value";
  }

  renderCell(columnIndex, value) {
    var colValue = this.formatColumnValue(columnIndex, value);
    return '<td>' + colValue + '</td>';
  }

  render(page) {
    let endPos = Math.min(this.panel.pageSize, this.table.rows.length);
    let startPos = 0;
    var html = "";

    for (var y = startPos; y < endPos; y++) {
      let row = this.table.rows[y];
      html += '<tr>';
      for (var i = 0; i < this.table.columns.length; i++) {
        html += this.renderCell(i, row[i]);
      }
      html += '</tr>';
    }

    return html;
  }
}
