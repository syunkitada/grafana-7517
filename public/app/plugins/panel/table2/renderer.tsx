// Libraries
import _ from 'lodash';
import moment from 'moment';
import React, { CSSProperties, ReactNode } from 'react';

import { sanitize } from 'app/core/utils/text';

// Types
import kbn from 'app/core/utils/kbn';
import { getValueFormat, getColorFromHexRgbOrName, GrafanaThemeType, InterpolateFunction } from '@grafana/ui';
import { Style } from './types';
import { SortedTableData } from './sortable';
import { Index } from 'react-virtualized';

type CellFormatter = (v: any, style: Style) => string;

interface ColumnInfo {
  header: string;
  accessor: string; // the field name
  style?: Style;
  hidden?: boolean;
  formatter: CellFormatter;
  filterable?: boolean;
}

export class TableRenderer {
  isUTC: false; // TODO? get UTC from props?

  columns: ColumnInfo[];
  colorState: any;
  theme?: GrafanaThemeType;

  constructor(
    styles: Style[],
    data: SortedTableData,
    private rowGetter: (info: Index) => any[], // matches the table rowGetter
    private replaceVariables: InterpolateFunction
  ) {
    this.colorState = {};

    if (!data) {
      this.columns = [];
      return;
    }

    this.columns = data.getInfo().map((col, index) => {
      let title = col.text;
      let style: Style = null;

      for (let i = 0; i < styles.length; i++) {
        const s = styles[i];
        const regex = kbn.stringToJsRegex(s.pattern);
        if (title.match(regex)) {
          style = s;
          if (s.alias) {
            title = title.replace(regex, s.alias);
          }
          break;
        }
      }

      return {
        header: title,
        accessor: col.text, // unique?
        style: style,
        formatter: this.createColumnFormatter(style, col),
      };
    });
  }

  setTheme(theme: GrafanaThemeType) {
    this.theme = theme;
  }

  getColorForValue(value, style: Style) {
    if (!style.thresholds) {
      return null;
    }
    for (let i = style.thresholds.length; i > 0; i--) {
      if (value >= style.thresholds[i - 1]) {
        return getColorFromHexRgbOrName(style.colors[i], this.theme);
      }
    }
    return getColorFromHexRgbOrName(_.first(style.colors), this.theme);
  }

  defaultCellFormatter(v: any, style: Style): string {
    if (v === null || v === void 0 || v === undefined) {
      return '';
    }

    if (_.isArray(v)) {
      v = v.join(', ');
    }

    if (style && style.sanitize) {
      return sanitize(v);
    } else {
      return _.escape(v);
    }
  }

  createColumnFormatter(style: Style, header: any): CellFormatter {
    if (!style) {
      return this.defaultCellFormatter;
    }

    if (style.type === 'hidden') {
      return v => {
        return undefined;
      };
    }

    if (style.type === 'date') {
      return v => {
        if (v === undefined || v === null) {
          return '-';
        }

        if (_.isArray(v)) {
          v = v[0];
        }
        let date = moment(v);
        if (this.isUTC) {
          date = date.utc();
        }
        return date.format(style.dateFormat);
      };
    }

    if (style.type === 'string') {
      return v => {
        if (_.isArray(v)) {
          v = v.join(', ');
        }

        const mappingType = style.mappingType || 0;

        if (mappingType === 1 && style.valueMaps) {
          for (let i = 0; i < style.valueMaps.length; i++) {
            const map = style.valueMaps[i];

            if (v === null) {
              if (map.value === 'null') {
                return map.text;
              }
              continue;
            }

            // Allow both numeric and string values to be mapped
            if ((!_.isString(v) && Number(map.value) === Number(v)) || map.value === v) {
              this.setColorState(v, style);
              return this.defaultCellFormatter(map.text, style);
            }
          }
        }

        if (mappingType === 2 && style.rangeMaps) {
          for (let i = 0; i < style.rangeMaps.length; i++) {
            const map = style.rangeMaps[i];

            if (v === null) {
              if (map.from === 'null' && map.to === 'null') {
                return map.text;
              }
              continue;
            }

            if (Number(map.from) <= Number(v) && Number(map.to) >= Number(v)) {
              this.setColorState(v, style);
              return this.defaultCellFormatter(map.text, style);
            }
          }
        }

        if (v === null || v === void 0) {
          return '-';
        }

        this.setColorState(v, style);
        return this.defaultCellFormatter(v, style);
      };
    }

    if (style.type === 'number') {
      const valueFormatter = getValueFormat(style.unit || header.unit);

      return v => {
        if (v === null || v === void 0) {
          return '-';
        }

        if (_.isString(v) || _.isArray(v)) {
          return this.defaultCellFormatter(v, style);
        }

        this.setColorState(v, style);
        return valueFormatter(v, style.decimals, null);
      };
    }

    return value => {
      return this.defaultCellFormatter(value, style);
    };
  }

  setColorState(value: any, style: Style) {
    if (!style.colorMode) {
      return;
    }

    if (value === null || value === void 0 || _.isArray(value)) {
      return;
    }

    if (_.isNaN(value)) {
      return;
    }
    const numericValue = Number(value);
    this.colorState[style.colorMode] = this.getColorForValue(numericValue, style);
  }

  renderRowVariables(rowIndex: number) {
    const scopedVars = {};
    const row = this.rowGetter({ index: rowIndex });
    for (let i = 0; i < row.length; i++) {
      scopedVars[`__cell_${i}`] = { value: row[i] };
    }
    return scopedVars;
  }

  renderCell(columnIndex: number, rowIndex: number, value: any): ReactNode {
    const column = this.columns[columnIndex];
    if (column.formatter) {
      value = column.formatter(value, column.style);
    }

    const style: CSSProperties = {};
    const cellClasses = [];
    let cellClass = '';

    if (this.colorState.cell) {
      style.backgroundColor = this.colorState.cell;
      style.color = 'white';
      this.colorState.cell = null;
    } else if (this.colorState.value) {
      style.color = this.colorState.value;
      this.colorState.value = null;
    }

    if (value === undefined) {
      style.display = 'none';
      column.hidden = true;
    } else {
      column.hidden = false;
    }

    if (column.style && column.style.preserveFormat) {
      cellClasses.push('table-panel-cell-pre');
    }

    let columnHtml: JSX.Element;
    if (column.style && column.style.link) {
      // Render cell as link
      const scopedVars = this.renderRowVariables(rowIndex);
      scopedVars['__cell'] = { value: value };

      const cellLink = this.replaceVariables(column.style.linkUrl, scopedVars, encodeURIComponent);
      const cellLinkTooltip = this.replaceVariables(column.style.linkTooltip, scopedVars);
      const cellTarget = column.style.linkTargetBlank ? '_blank' : '';

      cellClasses.push('table-panel-cell-link');
      columnHtml = (
        <a
          href={cellLink}
          target={cellTarget}
          data-link-tooltip
          data-original-title={cellLinkTooltip}
          data-placement="right"
        >
          {value}
        </a>
      );
    } else {
      columnHtml = <span>{value}</span>;
    }

    let filterLink: JSX.Element;
    if (column.filterable) {
      cellClasses.push('table-panel-cell-filterable');
      filterLink = (
        <span>
          <a
            className="table-panel-filter-link"
            data-link-tooltip
            data-original-title="Filter out value"
            data-placement="bottom"
            data-row={rowIndex}
            data-column={columnIndex}
            data-operator="!="
          >
            <i className="fa fa-search-minus" />
          </a>
          <a
            className="table-panel-filter-link"
            data-link-tooltip
            data-original-title="Filter for value"
            data-placement="bottom"
            data-row={rowIndex}
            data-column={columnIndex}
            data-operator="="
          >
            <i className="fa fa-search-plus" />
          </a>
        </span>
      );
    }

    if (cellClasses.length) {
      cellClass = cellClasses.join(' ');
    }

    style.width = '100%';
    style.height = '100%';
    columnHtml = (
      <div className={cellClass} style={style}>
        {columnHtml}
        {filterLink}
      </div>
    );
    return columnHtml;
  }
}
