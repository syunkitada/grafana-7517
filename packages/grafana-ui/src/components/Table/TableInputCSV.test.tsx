import React from 'react';

import renderer from 'react-test-renderer';
import TableInputCSV, { ParseResults } from './TableInputCSV';

describe('TableInputCSV', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <TableInputCSV
          width={100}
          height={100}
          onTableParsed={(results: ParseResults) => {
            console.log('GOT', results);
          }}
        />
      )
      .toJSON();
    //expect(tree).toMatchSnapshot();
    expect(tree).toBeDefined();
  });
});
