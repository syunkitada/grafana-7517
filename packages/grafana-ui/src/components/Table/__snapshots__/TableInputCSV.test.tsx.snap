// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableInputCSV renders correctly 1`] = `
<div>
  <textarea
    onBlur={[Function]}
    onChange={[Function]}
    value=""
  />
</div>
`;
